components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -0,0 +1,325 @@
+ * </code>
+ *
+ */
+@Suppress("TooManyFunctions")
+class AutoPushFeature(
+    private val context: Context,
+    private val service: PushService,
+    config: PushConfig,
+    coroutineContext: CoroutineContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher(),
+    private val connection: PushConnection = RustPushConnection(
+        senderId = config.senderId,
+        serverHost = config.serverHost,
+        socketProtocol = config.protocol,
+        serviceType = config.serviceType,
+        databasePath = File(context.filesDir, DB_NAME).canonicalPath
+    )
+) : PushProcessor {
+
+    private val logger = Logger("AutoPushFeature")
+    private val subscriptionObservers: Observable<PushSubscriptionObserver> = ObserverRegistry()
+    private val messageObserverBus: Bus<PushType, String> = MessageBus()
+    private val prefToken = preferences(context).getString(PREF_TOKEN, null)
+
+    internal var job: Job = SupervisorJob()
+    private val scope = CoroutineScope(coroutineContext) + job
+
+    init {
+        // If we have a token, initialize the rust component first.
+        prefToken?.let { token ->
+            runBlocking(coroutineContext) {
+                connection.updateToken(token)
+            }
+        }
+    }
+
+    /**
+     * Starts the push service provided.
+     */
+    override fun initialize() { service.start(context) }
+
+    /**
+     * Register to receive push subscriptions when requested or when they have been re-registered.
+     */
+    fun registerForSubscriptions(
+        observer: PushSubscriptionObserver,
+        owner: LifecycleOwner,
+        autoPause: Boolean
+    ) = subscriptionObservers.register(observer, owner, autoPause)
+
+    /**
+     * Register to receive push messages for the associated [PushType].
+     */
+    fun registerForPushMessages(
+        type: PushType,
+        observer: Bus.Observer<PushType, String>,
+        owner: LifecycleOwner,
+        autoPause: Boolean
+    ) = messageObserverBus.register(type, observer, owner, autoPause)
+
+    /**
+     * Un-subscribes from all push message channels. This should only be done on an account logout or app data deletion.
+     */
+    override fun shutdown() {
+        service.stop()
+
+        DeliveryManager.with(connection) {
+            scope.launch {
+                // TODO replace with unsubscribeAll API when available
+                PushType.values().forEach { type ->
+                    unsubscribe(type.toChannelId())
+                }
+                job.cancel()
+            }
+        }
+    }
+
+    /**
+     * New registration tokens are received and sent to the Autopush server which also performs subscriptions for
+     * each push type and notifies the subscribers.
+     */
+    override fun onNewToken(newToken: String) {
+        scope.launchAndTry {
+            connection.updateToken(newToken)
+
+            // Subscribe all only if this is the first time.
+            if (prefToken.isNullOrEmpty()) {
```

Comment:
who can modify `prefToken`? is this thread-safe?

Reply:
Only the feature modifies it. SharedPrefs are thread safe (not process safe).